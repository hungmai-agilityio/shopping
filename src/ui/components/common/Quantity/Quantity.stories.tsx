import type { Meta, StoryObj } from '@storybook/react';

// Constants
import { navItems } from '@/constants';
import Quantity from '@/ui/components/common/Quantity';

// Components

const meta: Meta<typeof Quantity> = {
  component: Quantity,
  tags: ['autodocs'],

  parameters: {
    layout: 'centered',
    nextjs: {
      appDirectory: true
    }
  },

  argTypes: {
    value: {
      description: 'value Quantity'
    },
    isBorder: {
      description: 'set border'
    }
  }
};

export default meta;

type Story = StoryObj<typeof Quantity>;

export const Default: Story = {
  args: {
    value: 2
  }
};

export const Second: Story = {
  args: {
    value: 2,
    isBorder: true
  }
};
