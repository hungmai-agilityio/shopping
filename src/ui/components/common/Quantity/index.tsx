'use client';

import React, { memo, useState } from 'react';
import { clsx } from 'clsx';

// Components
import Icon from '@/ui/components/common/Icon';
import { mada } from '@/constants';

interface QuantityProps {
  value: number;
  isBorder?: boolean;
  styles?: string;
  onQuantityChange?: (newQuantity: number) => void;
}

const Quantity = memo(
  ({ value = 1, isBorder, styles, onQuantityChange }: QuantityProps) => {
    const [quantity, setQuantity] = useState<number>(value);

    const handDecrease = () => {
      if (quantity > 1) {
        const newQuantity = quantity - 1;
        setQuantity(newQuantity);
        onQuantityChange?.(newQuantity);
      }
    };

    const handleIncrease = () => {
      const newQuantity = quantity + 1;
      setQuantity(newQuantity);
      onQuantityChange?.(newQuantity);
    };

    return (
      <div
        className={clsx(
          'flex items-center gap-4 font-bold bg-white w-36 h-12 justify-center',
          { 'border border-dark': isBorder },
          mada.className,
          styles
        )}
      >
        <Icon
          src="/minus.svg"
          alt="minus icon"
          width={24}
          height={24}
          onClick={handDecrease}
          isDisabled={quantity === 1}
          priority
        />
        <span className="px-2">{quantity}</span>
        <Icon
          src="/plus.svg"
          alt="plus icon"
          width={24}
          height={24}
          onClick={handleIncrease}
          priority
        />
      </div>
    );
  }
);

export default Quantity;
