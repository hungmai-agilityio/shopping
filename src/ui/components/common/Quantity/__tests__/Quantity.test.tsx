import Quantity from '@/ui/components/common/Quantity';
import { fireEvent, render, screen } from '@testing-library/react';

describe('Quantity component', () => {
  beforeEach(() => {
    jest.fn().mockClear();
  });

  it('render default props', () => {
    render(<Quantity value={3} />);
    expect(screen.getByText(3)).toBeInTheDocument();
  });

  it('handled when clicking on "minus"', () => {
    render(<Quantity value={3} />);
    const minus = screen.getByAltText('minus icon');
    fireEvent.click(minus);
    expect(screen.getByText(2)).toBeInTheDocument();
  });

  it('handled when clicking on "plus"', () => {
    render(<Quantity value={3} />);
    const plus = screen.getByAltText('plus icon');
    fireEvent.click(plus);
    expect(screen.getByText(4)).toBeInTheDocument();
  });

  it('should be render match to snapshot', () => {
    const { container } = render(<Quantity value={3} />);

    expect(container).toMatchSnapshot();
  });
});
