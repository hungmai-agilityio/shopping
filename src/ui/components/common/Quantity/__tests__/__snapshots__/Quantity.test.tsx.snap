// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quantity component should be render match to snapshot 1`] = `
<div>
  <div
    class="flex items-center gap-4 font-bold bg-white w-36 h-12 justify-center className"
  >
    <div
      class="relative inline-flex items-center justify-center cursor-pointer"
    >
      <img
        alt="minus icon"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="24"
        src="/minus.svg"
        style="color: transparent;"
        width="24"
      />
    </div>
    <span
      class="px-2"
    >
      3
    </span>
    <div
      class="relative inline-flex items-center justify-center cursor-pointer"
    >
      <img
        alt="plus icon"
        data-nimg="1"
        decoding="async"
        fetchpriority="high"
        height="24"
        src="/plus.svg"
        style="color: transparent;"
        width="24"
      />
    </div>
  </div>
</div>
`;
